src/dali/abstract_ccxt_pair_converter_plugin.py

```diff
@@ -0,0 +1,1027 @@
+            elif hop_data.exchange == _ALIAS:
+                hop_bar = current_graph.get_alias_bar(hop_data.from_asset, hop_data.to_asset, timestamp)
+            else:
+                hop_bar = self.find_historical_bar(hop_data.from_asset, hop_data.to_asset, timestamp, hop_data.exchange)
+
+            if hop_bar is not None:
+                # Replacing an immutable attribute
+                conversion_route[i] = conversion_route[i]._replace(historical_data=hop_bar)
+            else:
+                self._logger.debug(
+                    """No pricing data found for hop. This could be caused by airdropped
+                    coins that do not have a market yet. Market - %s%s, Timestamp - %s, Exchange - %s""",
+                    hop_data.from_asset,
+                    hop_data.to_asset,
+                    timestamp,
+                    hop_data.exchange,
+                )
+            if result is not None:
+                # TO BE IMPLEMENTED - override Historical Bar * to multiply two bars?
+                result = HistoricalBar(
+                    duration=max(result.duration, hop_bar.duration),  # type: ignore
+                    timestamp=timestamp,
+                    open=(result.open * hop_bar.open),  # type: ignore
+                    high=(result.high * hop_bar.high),  # type: ignore
+                    low=(result.low * hop_bar.low),  # type: ignore
+                    close=(result.close * hop_bar.close),  # type: ignore
+                    volume=(result.volume + hop_bar.volume),  # type: ignore
+                )
+            else:
+                result = hop_bar
+
+        return result
+
+    def find_historical_bar(self, from_asset: str, to_asset: str, timestamp: datetime, exchange: str) -> Optional[HistoricalBar]:
+        key: AssetPairAndTimestamp = AssetPairAndTimestamp(timestamp, from_asset, to_asset, exchange)
+        historical_bar: Optional[HistoricalBar] = self._get_bar_from_cache(key)
+
+        if historical_bar is not None:
+            self._logger.debug("Retrieved cache for %s/%s->%s for %s", timestamp, from_asset, to_asset, exchange)
+            return historical_bar
+
+        historical_bars: Optional[List[HistoricalBar]] = self.find_historical_bars(from_asset, to_asset, timestamp, exchange)
+
+        if historical_bars:
+            returned_bar: HistoricalBar = historical_bars[0]
+            if (timestamp - returned_bar.timestamp).total_seconds() > _TIME_GRANULARITY_STRING_TO_SECONDS[_ONE_DAY]:
+                raise RP2ValueError(
+                    "Internal error: The time difference between the requested and returned timestamps is "
+                    "greater than a day. The graph probably hasn't been optimized."
+                )
+            self._add_bar_to_cache(key=key, historical_bar=returned_bar)
+            return returned_bar
+        return None
+
+    def find_historical_bars(
```

Comment:
Opt/nit: this function is very large and could use some refactoring: consider breaking it into smaller bits in the future (you don't have to do it now, just keep it on your radar).

Reply:
Yeah 😄 There is a lot that needs to be refactored. I think I will spend more time on the core after this. :)